### ISSUE TITLE
Can't create job in custom namespace

### ISSUE THREAD

**Issue by Dmitry-Erokhin** [doriordan/skuber#192]
TL;DR: Attempt of creation job from application running in cluster lead to _"the namespace of the provided object does not match the namespace sent on the request"_

Job resource created by `val job = Job("gtd").withTemplate(template)` lead to Job metadata having hardcoded `default` namespace. Attempt to create such resource (from the app running on the pod in custom namespace) leads to
```
Failed to create job: skuber.api.client.package$K8SException: Status(v1,Status,ListMeta(,),Some(Failure),Some(the namespace of the provided object does not match the namespace sent on the request),Some(BadRequest),None,Some(400))
```
in logs

Temporary applied dirty workaround:

```
val badJob = Job("gtd").withTemplate(template)
val properJob = badJob.copy(metadata = badJob.metadata.copy(namespace = "custom_namespace"))
```

Am I missing something obvious?

**Comment by ngortheone** [comment]
It looks like a bug to me too. Here is a small self-contained example to reproduce with `Quickstart.sc` script: 
`amm -p ./Quickstart.sc`

```scala
import scala.concurrent.Await
import scala.concurrent.duration._

// We definitely have something in kube-system, so lets use it as example
val podsInKubeSystemReq = k8s.listInNamespace[PodList]("kube-system").map( p => p.items)
val podsInKubeSystem    = Await.result(podsInKubeSystemReq, 5.seconds)
val somePod             = podsInKubeSystem.head 
val someMetadata        = somePod.metadata
val someSpec            = somePod.spec

val someTemplate = Pod.Template.Spec(someMetadata, someSpec)
// just to make sure that template spec has correct namespace 
// @ someTemplate.metadata.namespace
// res18: String = "kube-system"

// Now lets create a job using example above
import skuber.batch.Job

val job = Job("gtd").withTemplate(someTemplate)
// @ job.namespace
// res21: String = "default"

//@ job.metadata.namespace
// res22: String = "default"

assert(job.namespace == someTemplate.metadata.namespace)
// or  assert(job.metadata.namespace == someTemplate.metadata.namespace)   --same story
```

I will try to investigate why this is so

**Comment by ngortheone** [comment]
@Dmitry-Erokhin 
So I believe this behavior is intentional but nevertheless some confusion exists and definitely skuber currently lacks some convenience methods. I will try to clarify what is going on.

## What happens
It all starts here:
As we can see from Job definition `withTemplate` does __not__ copy  metadata portion of the Template
```scala
 def withTemplate(template: Pod.Template.Spec) = this.copy(spec=Some(copySpec.copy(template=Some(template))))
```
And this is not a bug. To understand why we need to refer to the [official docs](https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/#writing-a-job-spec).  As it reads - all kubernetes objects need `apiVersion`, `kind`, and `metadata` fields. And `metadata` field is the one that can contain optional namespace field (see docs about [ObjectMeta](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#objectmeta-v1-meta))

And then there is `spec` section. This is the actual definition of the Job. According to the docs:
>The `.spec.template` is the only required field of the .spec.
>The `.spec.template` is a pod template. It has exactly the same schema as a pod, except it is nested and does not have an apiVersion or kind.

And this is reflected by Skuber.
```scala
case class Job(val kind: String ="Job",
               override val apiVersion: String = batchAPIVersion,
               val metadata: ObjectMeta = ObjectMeta(),
               spec: Option[Job.Spec] = None,
               status: Option[Job.Status] = None) extends ObjectResource
```

So when you do `.withTemplate` you create a Job with spec containing specified Pod template. You do not modify Job's metadata. 

## Confusion
### Job metadata vs Spec.Template metadata
The confusion arises from the fact that Pod Template also has its own metadata. Consider the following example. It is perfectly valid job definition:

```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: countdown
  namespace: mynamespace
spec:
  template:
    metadata:
      name: countdown
    spec:
      containers:
      - name: counter
        image: centos:7
        command:
         - "bin/bash"
         - "-c"
         - "for i in 9 8 7 6 5 4 3 2 1 ; do echo $i ; done"
      restartPolicy: Never
```
Notice that there are 2 `metadata` fields.  One is `.metadata` -describes metadata of the Job  and second one `.spec.template.metadata` - metadata of the template. I encourage you to play with this yaml file, modifying both metadata fields and feeding it to `kubectl`. You will notices that if you specify  namespace in `.spec.template.metadata` - it will be ignored, simply because template itself is not a namespaced object.

Here is some proof. Despite I specified `.spec.template.metadata` the output of `kubectl describe job` does not contain metadata field in spec
```
Name:           countdown3
Namespace:      default
Selector:       controller-uid=c7adac86-a037-11e8-912e-080027a3bc93
Labels:         controller-uid=c7adac86-a037-11e8-912e-080027a3bc93
                job-name=countdown3
Annotations:    <none>
Parallelism:    1
Completions:    1
Start Time:     Tue, 14 Aug 2018 23:03:26 -0400
Pods Statuses:  0 Running / 1 Succeeded / 0 Failed
Pod Template:
  Labels:  controller-uid=c7adac86-a037-11e8-912e-080027a3bc93
           job-name=countdown3
  Containers:
 ...
```
### Why `default` namespace?
It is very simple. Empty namespace field is treated as default namespace.
See `namespace` in [ObjectMeta](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#objectmeta-v1-meta) 

And when you create a Job it gets empty metadata:
- when you create with constructor `val metadata: ObjectMeta = ObjectMeta(),`
- when you create with companion object apply: `def apply(name: String) = new Job(metadata=ObjectMeta(name=name))`

And this leads to the next part.

## Where Skuber could be better:
@doriordan 
Currently there are 2 ways to specify the namespace where resource is to be created is
- Hackery, as described by @Dmitry-Erokhin 
`val properJob = badJob.copy(metadata = badJob.metadata.copy(namespace = "custom_namespace"))`
- Through manipulations with Configuration. The idea is to create separate client object with different namespace, like: 
```scala
val k8s = k8sInit

k8s.usingNamespace("mycustomnamespace").create(job)
```

We could probably add convenience method to `RequestContext` , something similar to `k8s.listInNamespace`, maybe `createInNamespace` as a shortcut to `usingNamespace.create`
And yes resources should have better convenient methods for metadata manipulation. The first idea that comes to mind is to extend existing "builder" pattern and add methods like `withNamespace`

Sorry for long post, I hope I was able to bring some clarity.

**Comment by ngortheone** [comment]
@doriordan 
Separately, this should be labled as not a bug. And definitely worth adding to FAQ section

**Comment by agemooij** [comment]
Those extra `withNamespace` and `createInNamespace` would indeed be very handy. 

We are indeed currently switching the namespace in the request context when we need to create resources in another namespace and that is not very convenient.

**Comment by Dmitry-Erokhin** [comment]
From what I saw in the code, the source of the problem is actually mismatch between resource data and url it's posted to. Skuber is trying to apply resource with "default" namespace in metadata to address which path is including not default one:

Essentially it is equals to this yaml
```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: countdown
  namespace: default #<<<<<<<<<<<<=========== note this one
spec:
  #some spec...
```
posted to `https://kubernetes.default/apis/batch/v1/namespaces/custom/jobs`

note "namespaces/**custom**"


And if my application is running in `custom` namespace this would happened just because 

```scala
 object Job {
  //...
  def apply(name: String) = new Job(metadata=ObjectMeta(name=name))`
 //...
}
```
and

```scala
  case class ObjectMeta(
    //...
    namespace: String = "default",
   //...
)
```

**Comment by doriordan** [comment]
I agree with the above, the current process for creating resources in different namespaces with skuber is non-intuitive, not documented clearly, and not very nice.
I like the idea of `createInNamespace` and `withNamespace` - in conjunction with using the existing `usingNamespace` for non-create ops plus a section in the docs on programming with multiple namespaces I think this would meet most use cases in a better way.

**Comment by doriordan** [comment]
Another option instead of `createInNamespace` is to add a flag to RequestContext which tells it whether to override the namespace in the object with that of the request context. By default we could set that to `false` (current behaviour) but when `usingNamepace` is called it sets the flag to `true` in the returned request context so the following should work:
```
val template=jobTemplate.withNamespace("custom")
val job = Job("gtd").withTemplate(template)
k8s.usingNamespace("custom").create(job)
```

**Comment by agemooij** [comment]
The I would prefer that the default is `true` and no explicit call is needed. 

Specifying your target namespace twice is kind of weird anyway. If your resource specifies its target namespace (using `'withNamespace`), then the most intuitive behavior would be for Skuber to respect that and create the resource in that namespace, right?

**Comment by Dmitry-Erokhin** [comment]
Or perhaps it will be logical that pod running in namespace A and creating resource via skuber should create it in same namespace by default. Creating resource in the namespace different from those on which pod is running is a more advanced case and can be configured explicitly.

IE:
1. Pod running in namespace A create a job in same namespace
```scala
val job = Job("gtd").withTemplate(jobTemplate) //No namespace specification
k8s.create(job)
```

2. Pod running in namespace A create a job in namespace B
```scala
val template=jobTemplate
val job = Job("gtd").withTemplate(jobTemplate).inNamepace("B") //Specify ns on job level
k8s.create(job)
```

**Comment by ngortheone** [comment]
@Dmitry-Erokhin unfortunately I have to disagree. As I explained above Pod Template is not a namespace object. 
```
.withTemplate(jobTemplate)  <== this does not carry namespace information!
```
Creating a pod/job from a template of another pod should not affect its namespace - this is the way kubectl and client-go works. Changing this will only create confusion. If you want - you can clone existing pod/job object if you want it to have the same namespace. 

Unless someone has already started - I can start putting together a PR to add discussed functionality.

**Comment by agemooij** [comment]
Ah, yes. it should be something like:

```scala
val job = Job("gtd").withTemplate(jobTemplate).withNamespace("foo")
k8s.create(job)
```

It shouldn't matter in which namespace the calling code might be running IMHO. If I've explicitly created a `Job` with a specific `namespace` (metadata) attribute, then I would expect Skuber to "make it so" by mapping that `create` to the endpoint in the proper namespace.

Yes, if you literally translate the actual API you would have to specify that namespace in two places, e.g in the URI/path and in the resource object you `POST`, but I would argue that Skuber exists to hide such inconveniences.

**Comment by doriordan** [comment]
I think part of the issue may be that we don't distinguish between an object where the namespace has not been set, and an object where the namespace has been set to the default namespace.

If we could distinguish between those two cases clearly, then if the namespace has not been set in the object metadata we could always just set it to that of the request context when `create` is called. If the namespace has been explicitly set on the object then we use that.

If we always rewrite it the concern would be that objects may end up being created on namespaces other than that set by the user on the object, which may be surprising especially if they are using default config with no explicit namespace.

At the moment `metadata.namespace` is a `String` rather than an `Option[String]` and is set to "default" by default because that is the Kubernetes default. However perhaps it should be set to an empty string by default, and then that indicates it has not been explicitly set by the user as opposed to having been set explicitly to "default". My concern is whether this might break some existing code.

**Comment by ngortheone** [comment]
This should be fixed by now with #203 and #211

**Comment by doriordan** [comment]
Closing this as fixed per above
